core/src/verified_vote_packets.rs

```diff
@@ -208,95 +208,94 @@ impl VerifiedVotePackets {
         vote_packets_receiver: &VerifiedLabelVotePacketsReceiver,
         would_be_leader: bool,
     ) -> Result<()> {
+        if !would_be_leader {
+            return Ok(());
+        }
+
         use SingleValidatorVotes::*;
         const RECV_TIMEOUT: Duration = Duration::from_millis(200);
         let vote_packets = vote_packets_receiver.recv_timeout(RECV_TIMEOUT)?;
         let vote_packets = std::iter::once(vote_packets).chain(vote_packets_receiver.try_iter());
 
         for gossip_votes in vote_packets {
-            if would_be_leader {
```

Comment:
we can't early return here because we still want to clear the receiver of votes when not close to our leader slot. Otherwise when `would_be_leader` is true, we will have many stale votes to process.

Reply:
got it, will place it right before a for loop.

Reply:
Good point about wanting to keep the channel pruned. If we're not the leader, I think we could avoid blocking with `recv_timeout()` and just do something like this at the top?
```rust
if !would_be_leader {
    let _vote_packets_to_discard: Vec<_> = vote_packets_receiver.try_iter().collect();
    return Ok(());
}
```

Reply:
i guess this also works

Reply:
Although on second though, we might need to look at the caller (`ClusterInfoVoteListener`); if we're waiting up to 200ms to get packets out of the channel, this might implicitly be limiting how often we grab the PoH recorder lock in the caller (this block):
https://github.com/anza-xyz/agave/blob/d7b22e2ae35e0336b1a1ada78adaf653b8a680eb/core/src/cluster_info_vote_listener.rs#L397-L420

Reply:
@steviez  Yes, but `.recv_timeout(RECV_TIMEOUT)` will not necessarily wait for 200ms, and will return votes instantly if there are on a channel. Still, it didn't block a PohRecorder lock, so I think your suggestion should work.

Reply:
@steviez looks like `ClusterInfoVoteListener` could be optimized to not grab the poh lock that often. 
It already is grabbing the lock every 10ms 😬 which seems unnecessary until we are close to our leader slots.
Similarly it grabs the lock every iteration for `would_be_leader` even though once this is `true` we can stop checking for the next `3 * 512` slots.

Reply:
> Yes, but .recv_timeout(RECV_TIMEOUT) will not necessarily wait for 200ms, and will return votes instantly if there are on a channel

Correct - this is why I prefaced my statement with "In the case where the channel is empty"

> looks like ClusterInfoVoteListener could be optimized to not grab the poh lock that often.

💯 

> Similarly it grabs the lock every iteration for would_be_leader

💀

> It already is grabbing the lock every 10ms 😬 which seems unnecessary until we are close to our leader slots.

I think you're right. With `PohRecorder::would_be_leader()` check, we are seeing if we'll be leader in the next `3 * slot_hashes::MAX_ENTRIES == 3 * 512 slots`. This means `would_be_leader` will toggle true when we have a leader slot in the next:
```
3 * 512 slots * 400ms / slot = 614s = 10.25 minutes
``` 
Beginning to examine votes 10.25 minutes in advance seems excessive. In any case, we don't need to check every 10ms if we start processing votes 10 minutes ahead of time. I think both of these numbers could potentially be tuned

> Similarly it grabs the lock every iteration for would_be_leader even though once this is true we can stop checking for the next 3 * 512 slots.

Agreed

--

Some good conversation came out of this. For the scope of this PR, I think @kubanemil moving the early return after draining the channel (while keeping `recv_timeout()`) is sufficient, and we can make any other desired changes in a separate PR. Sound good @AshwinSekar ?

Reply:
sounds good to me.